### PR TITLE
#168781628 - Implementation of notifications Panel

### PR DESCRIPTION
**Pull request by nignanthomas**
#### What does this PR do?
Implementation of notifications Panel.

#### Description of Task to be completed?
- Create actions/reducers to get all notifications, mark them and switch notifications
- Create component to display all notifications and button to mark as read/unread
- Create dropdown to display 10 most recent notifications, mark all as read, see more

#### How should this be manually tested?
- Clone ft-notifications-pane-168781628 branch
- run `npm i`
- run `npm run dev`
- Click on `Notifications` dropdown to view most recent notifications and/or view all notifications
#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#168781628](https://www.pivotaltracker.com/story/show/168781628)

#### Screenshots (if appropriate)
Screen Shot 2019-12-28 at 19 14 19
Screen Shot 2019-12-28 at 19 14 29

#### Questions:
N/A

### CODE REVIEW

**Review by alainmateso — CHANGES_REQUESTED** [review]
Great Job Thomas. Kindly work on the following feedback:

- When the user has no notifications, it is better to tell them in the panel instead of bringing mark all as read and see more because if a user clicks on see more, it only comes an empty page

- I believe the notification counter(the number itself) should be white to facilitate readability

- When I click on a request from a notification, it should mark that notification as read. It does not make sense to me to click on a notification and I doesn't turn to read. That means the user will always have unread notifications even though in reality they are read.

**Review by ngireric123 — COMMENTED** [review]
Great Job @nignanthomas  Kindly try to increase test coverage

File: src/components/global/MenuComponent.js [hunk 1]
```diff
@@ -25,13 +27,34 @@ export class MenuComponent extends Component {
...
+    };
+
     render() {
       const { props } = this;
-      const { data, pathname } = props;
+      const { data, pathname, notifsData } = props;
       const urls = ['/login', '/register', '/forgotpassword', '/registered'];
       const displayMenu = !(urls.includes(pathname) || pathname.match(/resetpassword/) || pathname.match(/verify/));
+      let unreadNotifs = 0;
+      console.log('props.notifsData ===> ', props.notifsData);
```
Comment (nakiwuge) [inline src/components/global/MenuComponent.js]: remove console

File: src/components/global/MenuComponent.js [hunk 2]
```diff
@@ -25,13 +27,34 @@ export class MenuComponent extends Component {
...
+      if (notifsData) {
+        unreadNotifs = notifsData.filter((notif) => !notif.isRead).length;
+      }
+      console.log('unreadNotifs ===> ', unreadNotifs);
```
Comment (nakiwuge) [inline src/components/global/MenuComponent.js]: remove the console

File: src/components/pages/notifications/NotificationItem.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  Container, Row, Col, Button,
+} from 'react-bootstrap';
+import { Link } from 'react-router-dom';
+import PropTypes from 'prop-types';
+import moment from 'moment';
+import { connect } from 'react-redux';
+import { markOneNotifAction, getNotifsAction } from '../../../actions/notificationsActions';
+
+export class NotificationItem extends Component {
+  state = {
```
Comment (nakiwuge) [inline src/components/pages/notifications/NotificationItem.js:12]: Remove empty state and if you are not changing state make this a functional component

File: src/components/pages/notifications/NotificationItem.js [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+  };
+
+  markUnmark = async (notifId) => {
+    const { props } = this;
+    console.log('notifId ===> ', notifId);
```
Comment (nakiwuge) [inline src/components/pages/notifications/NotificationItem.js]: Remove this console

File: src/components/pages/notifications/Notifications.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+import NotificationItem from './NotificationItem';
+
+export class Notifications extends Component {
+  state = {
+    isLoading: false,
+  };
+
+  async componentDidMount() {
+    const { props } = this;
+    const { getNotifsAction } = props;
```
Comment (nakiwuge) [inline src/components/pages/notifications/Notifications.js:20]: Add a new line here

File: src/components/pages/notifications/Notifications.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    this.setState({ isLoading: true });
+    await getNotifsAction();
+    this.setState({ isLoading: false });
+  }
+
+  render() {
+    const { props, state: { isLoading } } = this;
+    const { notifsData } = props;
```
Comment (nakiwuge) [inline src/components/pages/notifications/Notifications.js:35]: Add a new line here

File: src/components/pages/profiles/UserProfile.js [hunk 7]
```diff
@@ -112,10 +118,27 @@ export class UserProfile extends Component {
     await props.UpdateUserProfile(form);
   };
 
+  switchNotif = async (switchParam) => {
+    const { props } = this;
+    this.setState({ isLoading: true });
+    await props.switchNotifAction(switchParam);
+    await props.GetUserProfile();
+    const { data } = this.props;
```
Comment (nakiwuge) [inline src/components/pages/profiles/UserProfile.js]: move the constants to the top

File: src/reducers/allNotifsReducer.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
+import { GET_ALL_NOTIF_SUCCESS, GET_ALL_NOTIF_ERROR } from '../actions/types';
+
+const initialState = {
+  notifsData: null,
+  notifsDataError: null,
+  // status: '',
```
Comment (nakiwuge) [inline src/reducers/allNotifsReducer.js]: remove commented code

File: src/reducers/allNotifsReducer.js [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+};
+
+export default (state = initialState, action) => {
+  switch (action.type) {
+    case GET_ALL_NOTIF_SUCCESS:
+      return {
+        ...state,
+        notifsData: action.payload.data,
+        notifsDataError: null,
+        // status: 'success',
```
Comment (nakiwuge) [inline src/reducers/allNotifsReducer.js]: remove commented code

File: src/reducers/allNotifsReducer.js [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+      };
+    case GET_ALL_NOTIF_ERROR:
+      return {
+        ...state,
+        notifsData: null,
+        notifsDataError: action.payload,
+        // status: 'error',
```
Comment (nakiwuge) [inline src/reducers/allNotifsReducer.js]: remove commented code

File: src/reducers/allNotifsReducer.js [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+      };
+    default:
+      return {
+        ...state,
+        // status: '',
```
Comment (nakiwuge) [inline src/reducers/allNotifsReducer.js]: Remove commented code

File: src/styles/components/menubar.scss [hunk 12]
```diff
@@ -3,7 +3,8 @@
 	flex-direction: row;
 	align-items: center;
 	font-size: 12px;
-	justify-content: space-between;
+	// justify-content: space-between;
```
Comment (nakiwuge) [inline src/styles/components/menubar.scss]: Remove commented code

File: src/styles/notifications.scss [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+  font-size: 12px;
+}
+
+.notif-number {
+  background-color: $googleRedDark;
+  color: $white;
+  padding: 0.2rem;
+  margin-top: 2rem;
+  margin-left: -1rem;
+  // border-radius: 100%;
```
Comment (nakiwuge) [inline src/styles/notifications.scss]: Remove commented code

**Review by nakiwuge — CHANGES_REQUESTED** [review]
Great work @nignanthomas I am impressed with your work however kindly work on the feedback below.
 - Give the `Mark all as read` a blue color so that it stands out.
- A user should be able to reverse the `Mark all as read` action (Marl all as unread)
- You should not show the `request id` in the notification just leave the word REQUEST
- The `modal/notifications panel` is not aligned with the `Notification tab` on the menu when the user has no notifications.
- The Mark as read on the notification page is not working as expected. as shown in the gif below.
- I have left other comments on the pr.

Dec-30-2019 19-41-25 no

**Review by nakiwuge — APPROVED** [review]

**Pull request merged**